Ingress/controllers/nginx-third-party/ssl/main.go

```diff
@@ -0,0 +1,180 @@
+		cert := firstFile(fmt.Sprintf("%v/%v", baseDir, file.Name()), ".crt")
+		key := strings.Replace(cert, ".crt", ".key", 1)
+
+		if _, err := os.Stat(cert); os.IsNotExist(err) {
+			continue
+		}
+
+		glog.Infof("\t%v", cert)
+
+		hosts, err := checkSSLCertificate(cert, key)
+		if err == nil {
+			sslCerts = append(sslCerts, Certificate{
+				Cert:  cert,
+				Key:   key,
+				Cname: hosts,
+				Valid: true,
+			})
+		} else {
+			glog.Errorf("error checking certificate: %v", err)
+		}
+	}
+
+	glog.Infof("ssl certificates found: %v", sslCerts)
+
+	return sslCerts
+}
+
+// checkSSLCertificate check if the certificate and key file are valid
+// returning the result of the validation and the list of hostnames
+// contained in the common name/s
+func checkSSLCertificate(certFile, keyFile string) ([]string, error) {
+	_, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return []string{}, err
+	}
+
+	pemCerts, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return []string{}, err
+	}
+
+	var block *pem.Block
+	block, pemCerts = pem.Decode(pemCerts)
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return []string{}, err
+	}
+
+	cn := []string{cert.Subject.CommonName}
+	if len(cert.DNSNames) > 0 {
+		cn = append(cn, cert.DNSNames...)
+	}
+
+	glog.V(2).Infof("DNS %v %v\n", cn, len(cn))
+	return cn, nil
+}
+
+func verifyHostname(certFile, host string) bool {
+	pemCerts, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return false
+	}
+
+	var block *pem.Block
+	block, pemCerts = pem.Decode(pemCerts)
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+
+	err = cert.VerifyHostname(host)
+	if err == nil {
+		return true
+	}
+
+	return false
+}
+
+// GetSSLHost checks if in one of the secrets that contains SSL
+// certificates could be used for the specified server name
+func GetSSLHost(serverName string, certs []Certificate) Certificate {
+	for _, sslCert := range certs {
+		if verifyHostname(sslCert.Cert, serverName) {
+			return sslCert
+		}
+	}
+
+	return Certificate{}
+}
+
+// SearchDHParamFile iterates all the secrets mounted inside the /etc/nginx-ssl directory
+// in order to find a file with the name dhparam.pem. If such file exists it will
+// returns the path. If not it just returns an empty string
+func SearchDHParamFile(baseDir string) string {
+	files, _ := ioutil.ReadDir(baseDir)
+	for _, file := range files {
+		if !file.IsDir() {
+			continue
+		}
+
+		dhPath := fmt.Sprintf("%v/%v/dhparam.pem", baseDir, file.Name())
+		if _, err := os.Stat(dhPath); err == nil {
+			glog.Infof("using file '%v' for parameter ssl_dhparam", dhPath)
+			return dhPath
+		}
+	}
+
+	glog.Warning("no file dhparam.pem found in secrets")
+	return ""
+}
+
+// firstFile returns the first file in the specified directory with the
+// extension
+func firstFile(baseDir, ext string) string {
+	var file string
+	filepath.Walk(baseDir, func(path string, f os.FileInfo, _ error) error {
```

Comment:
I think this will do a lexical walk, so the crt might change? if possible we should sort on timestamp, otherwise document it.